sources/rest_api/__init__.py

```diff
@@ -0,0 +1,768 @@
+    resource: ResourceConfig
+    incremental: Optional[IncrementalConfig]
+    method: str
+    paginator: Optional[PaginatorType]
+
+
+class RESTAPIConfigLegacy(TypedDict):
+    client: ClientConfig
+    endpoints: Dict[str, EndpointConfig]
+
+
+class ResolveConfig(NamedTuple):
+    resource_name: str
+    field_path: str
+
+
+class ResolvedParam(NamedTuple):
+    param_name: str
+    resolve_config: ResolveConfig
+
+
+class Endpoint(TypedDict, total=False):
+    path: str
+    method: str
+    params: Optional[Dict[str, Any]]
+    json: Optional[Dict[str, Any]]
+    paginator: Optional[PaginatorType]
+
+
+class EndpointResource(TypedDict, total=False):
+    name: TTableHintTemplate[str]
+    endpoint: Endpoint
+    write_disposition: TTableHintTemplate[TWriteDisposition]
+    parent: TTableHintTemplate[str]
+    columns: TTableHintTemplate[TTableSchemaColumns]
+    primary_key: TTableHintTemplate[TColumnNames]
+    merge_key: TTableHintTemplate[TColumnNames]
+    incremental: Incremental[Any]
+    # schema_contract: TTableHintTemplate[TSchemaContract]
+    table_format: TTableHintTemplate[TTableFormat]
+
+
+class RESTAPIConfig(TypedDict):
+    client: ClientConfig
+    resource_defaults: EndpointResource
+    resources: Dict[str, EndpointResource]
+
+
+def create_paginator(paginator_config):
+    if isinstance(paginator_config, BasePaginator):
+        return paginator_config
+    return PAGINATOR_MAP.get(paginator_config, lambda: None)()
+
+
+def create_auth(auth_config):
+    if isinstance(auth_config, AuthBase):
+        return auth_config
+    return BearerTokenAuth(auth_config.get("token")) if auth_config else None
+
+
+def make_client_config(config):
+    client_config = config.get("client", {})
+    return {
+        "base_url": client_config.get("base_url"),
+        "auth": create_auth(client_config.get("auth")),
+        "paginator": create_paginator(client_config.get("default_paginator")),
+    }
+
+
+def setup_incremental_object(request_params, incremental_config):
+    for key, value in request_params.items():
+        if isinstance(value, dlt.sources.incremental):
+            return value, key
+        if isinstance(value, dict) and value.get("type") == "incremental":
+            return (
+                dlt.sources.incremental(
+                    value.get("cursor_path"), initial_value=value.get("initial_value")
+                ),
+                key,
+            )
+
+    return setup_incremental_object_from_config(incremental_config)
+
+
+def setup_incremental_object_from_config(config):
+    return (
+        (
+            dlt.sources.incremental(
+                config.get("cursor_path"), initial_value=config.get("initial_value")
+            ),
+            config.get("param"),
+        )
+        if config
+        else (None, None)
+    )
+
+
+def make_parent_key_name(resource_name, field_name):
+    return f"_{resource_name}_{field_name}"
+
+
+@dlt.source
+def rest_api_source(config: RESTAPIConfig):
```

Comment:
do not decorate. use `dlt.source` as function. allow for the custom name, section, schema, max table nesting and other params that decorator accepts